Number_theory.hpp

```diff
@@ -0,0 +1,261 @@
+
+        decltype(a) x;
+
+        if (b % a == 0 && mod % a == 0)
+        {
+            x = (b / a) % (mod / a);
+            return x;
+        }
+
+        if (gcd(a, mod) != 1)
+        {
+            throw std::invalid_argument("a and mod must be coprime.");
+        }
+
+        auto inverse = modular_inverse(a, mod);
+        x = (b * inverse) % mod;
+        return x;
+    }
+
+    /*returns primes smaller than up_to*/
+    inline auto sieve_of_eratosthenes(std::size_t up_to)
+    {
+        std::vector<bool> is_prime(up_to + 1, true);
+        std::vector<size_t> primes;
+
+        for (size_t i = 2; i < up_to; i++)
+        {
+            if (is_prime[i])
+            {
+                for (size_t j = 2 * i; j < up_to; j += i)
+                {
+                    is_prime[j] = false;
+                }
+            }
+        }
+
+        for (size_t i = 2; i < up_to; i++)
+        {
+            if (is_prime[i])
+            {
+                primes.push_back(i);
+            }
+        }
+        return primes;
+    }
+
+    /*decomposes a number into powers of primes*/
+    inline auto decompose(std::integral auto value)
+    {
+        std::vector<std::pair<int, int>> primeFactors;
+
+        for (int i = 2; i <= value / i; ++i)
+        {
+            int count = 0;
+            while (value % i == 0)
+            {
+                ++count;
+                value /= i;
+            }
+            if (count > 0)
+            {
+                primeFactors.emplace_back(i, count);
+            }
+        }
+        if (value > 1)
+        {
+            primeFactors.emplace_back(value, 1);
+        }
+        return primeFactors;
+    }
+
+    /*calculates Euler�s Totient Function*/
+    inline auto euler_totient(std::integral auto value)
+    {
+        decltype(value) result = 0;
+        for (int i = 1; i <= value; i++) {
+            if (gcd(i, value) == 1) {
+                result++;
+            }
+        }
+        return result;
+    }
+
+    /*finds the largest power pow of prime such that prime^pow divides value*/
+    inline auto largest_power_of_prime_dividing_factorial(std::integral auto value, std::integral auto prime)
+    {
+        decltype(prime) pow = 0;
+        while (value % prime == 0)
+        {
+            pow++;
+            value /= prime;
+        }
+        return pow;
+    }
+
+    /*solves a system of linear equations as[i] * x = bs[i] mod mods[i].*/
+    template <std::integral I>
+    inline auto linear_congruence_solver(std::vector<I> as, std::vector<I> bs, std::vector<std::size_t> mods)
+    {
+        if (as.size() != bs.size() || bs.size() != mods.size())
+        {
+            throw std::invalid_argument("Input vectors must have the same size.");
+        }
+
+
+        for (size_t i = 0; i < as.size(); i++)
+        {
+
+            try
+            {
+                auto a_inverse = modular_inverse(as[i], mods[i]);
+                bs[i] = (bs[i] * a_inverse) % mods[i];

```

Comment:
exceptions as the name suggests are used ONLY when sth very exceptional occurs.... you can use exceptions logic to implement functions. I suggest after reading you code that you use std::expected + enum class error {invalid_argument, ... };